algorithms.py

```diff
@@ -825,6 +890,160 @@ def hudson_recombination_event(self, label, return_heads=False):
             ret = x, z
         return ret
 
+    def wiuf_geneconversion_within_event(self, label, return_heads=False):
+        """
+        Implements a gene conversion event that starts within a segment
+        """
+        h = random.randint(1, self.L[label].get_total())
+        # generate tracklength
+        tl = np.random.geometric(1/self.tracklength)
+        # Get the segment containing the h'th link
+        y = self.segments[self.L[label].find(h)]
+        k = y.right - self.L[label].get_cumulative_frequency(y.index) + h - 1
+        # check if the gene conversion falls between segments --> no effect
+        if y.left >= k+tl:
+            # print("noneffective GCI EVENT")
+            return None
+        self.num_gc_events += 1
+        x = y.prev
+        # both breaks are within the same segment
+        if k+tl < y.right:
+            if k <= y.left:
+                y.prev = None
+                z2 = self.alloc_segment(
+                    k+tl, y.right, y.node, y.population, x, y.next)
+                if x is not None:
+                    x.next = z2
+                    lhs = x.right
+                    lhs_tail = x
+                else:
+                    print("this should never be possible")
+                    lhs = z2.left
+                self.L[label].set_value(z2.index, z2.right - lhs)
+                if y.next is not None:
+                    y.next.prev = z2
+                y.next = None
+                y.right = k+tl
+                self.L[label].increment(y.index, k + tl - z2.right)
+                z = y
+            elif k > y.left:
+                z = self.alloc_segment(
+                    k, k+tl, y.node, y.population, None, None)
+                z2 = self.alloc_segment(
+                    k+tl, y.right, y.node, y.population, y, y.next)
+                if y.next is not None:
+                    y.next.prev = z2
+                y.next = z2
+                y.right = k
+                self.L[label].set_value(z2.index, z2.right - y.right)
+                self.L[label].increment(y.index, k - z2.right)
+                lhs_tail = y
+        # breaks are in separate segments
+        else:
+            # Get the segment y2 containing the end of the conversion tract
+            y2 = y
+            while y2 is not None and k+tl >= y2.right:
+                y2 = y2.next
+            # process left break
+            if k <= y.left:
```

Comment:
This if/else block looks very similar to the one above - could this logic be pulled out into a helper function?

Reply:
Good idea. I wrote a helper function cut_right_break to do so.